src/main/java/bio/terra/stairway/FlightMap.java

```diff
@@ -5,24 +5,86 @@
 import bio.terra.stairway.exception.JsonConversionException;
 import com.fasterxml.jackson.core.JsonProcessingException;
 import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.deser.std.StdDelegatingDeserializer;
+import com.fasterxml.jackson.databind.module.SimpleModule;
+import com.fasterxml.jackson.databind.ser.std.StdDelegatingSerializer;
+import com.fasterxml.jackson.databind.util.StdConverter;
+import com.google.common.annotations.VisibleForTesting;
 import java.io.IOException;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.stream.Collectors;
+import org.springframework.lang.Nullable;
 
 /**
  * FlightMap wraps a {@code HashMap<String, Object>} It provides a subset of the HashMap methods. It
  * localizes code that casts from Object to the target type. It provides a way to set the map to be
  * immutable.
  */
 public class FlightMap {
-  private Map<String, Object> map;
+  @VisibleForTesting
+  static final class ObjectContainer {
+    private final String data;
+    private final Object object;
+    private final FlightParameterSerializer serializer;
+
+    public ObjectContainer(String data) {
+      this.data = data;
+      this.object = null;
+      this.serializer = null;
+    }
+
+    public ObjectContainer(Object object, FlightParameterSerializer serializer) {
+      this.data = null;
+      this.object = object;
+      this.serializer = serializer;
+    }
+
+    public String getData() {
+      // Constructors ensure that one can only have valid data OR object+serializer.
+      return (data != null) ? data : serializer.serialize(object);
+    }
+
+    public <T> T getObject(FlightParameterDeserializer<T> deserializer) {
+      return (object != null) ? deserializer.safeCast(object) : deserializer.deserialize(data);
```

Comment:
Is immutability worth repeated deserialization?
I think it'd be better to store the resulting object.

Similarly, at each step boundary, we will serialize the working map, but won't save the serialization work in `data`.

Reply:
My first pass cached a reference to the object on deserialization in the call to `getObject()`, but that raised some ambiguity around what the right behavior would be at serialization time if the referenced object and the serialized data have diverged from one another.

It sounds like I took the exact wrong approach though; would the preferable approach be that the referenced object is the object-of-record and the serialized data member is only valid until the object gets populated?  If so, renaming `data` to `initialSerializedState` would make this clear when I change the behavior.

The implication would be that any changes to the referenced object between `put` or `get` and serialization will be what is recorded at serialization time.

Reply:
Yet another alternative (with some downsides) is to only keep the serialized form. When the object is put, we immediately serialize it. When it is gotten, the deserializer is provided and we make the object. We never try to keep track of the serializer in the mao. That would reduce the map to `<String, String>`

One benefit of that approach is the normal (no fail overs path) would break if the serdes was broken. These days we only find out about it when there is a failure and the deserialization fails. Not so great.

The bad part is we would spend more time doing deserialization. We already serialize everything at each step boundary.

Reply:
From a code complexity standpoint, this would be a big win, as it would eliminate a bunch of type erasure related messiness that arises from storing references as `Object`, and make the ser/des of the map simpler.  It also makes it perfectly clear that `get` and `put` are the only ways to effect changes to what gets serialized, and eliminates any implicit state changes due to reference mutability.

Does "serializing everything at each step boundary" imply that `toJson()` would be expected to be called more frequently per-parameter than `get()/put()`?  If so, it sounds like this might be an efficiency win too, as we'd only serialize once, and then reuse the serialized data.  We should also figure in that we'll be paying for a deserialization on the first `get()` anyway...

Reply:
In the current scheme, the input parameters are serialized at flight launch and are only deserialized again if the flight is queued, either directly or if there is a failover and the flight needs to be restarted. The working map is serialized at the end of each step is deserialized on flight restart.

Just keeping the serialized state in the map is probably a wash. There would be more deserialization and less serialization. Typically, working map entries are written once and then read one or more times.